backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt

```diff
@@ -0,0 +1,161 @@
+                appendln()
+                appendln("Included libraries:")
+                includedLibraries.forEach { appendln(it.libraryFile) }
+            }
+
+            configuration.report(CompilerMessageSeverity.STRONG_WARNING, message)
+        }
+    }
+
+    private class ExportedLibrariesReporter(configuration: CompilerConfiguration) : BaseReporter(configuration) {
+        override fun illegalKindMessage(kind: String, libraryName: String): String =
+            "$kind library $libraryName can't be exported with -Xexport-library"
+
+        override fun notIncludedLibraryMessageTitle(): String =
+            "Following libraries are specified to be exported with -Xexport-library, but not included to the build:"
+    }
+
+    private class SourceLibrariesReporter(configuration: CompilerConfiguration) : BaseReporter(configuration) {
+        override fun illegalKindMessage(kind: String, libraryName: String): String =
+            "$kind library $libraryName can't be used as a source library"
+
+        override fun notIncludedLibraryMessageTitle(): String =
+            "Following libraries are declared as source libraries with -Xsource-library, but not included to the build:"
+    }
+
+    private class CoveredLibraryReporter(configuration: CompilerConfiguration): BaseReporter(configuration) {
+        override fun illegalKindMessage(kind: String, libraryName: String): String =
+            "$kind library $libraryName can't be covered"
+
+        override fun notIncludedLibraryMessageTitle(): String =
+            "Following libraries are specified to be covered with -Xlibrary-to-cover, but not included to the build:"
+    }
+
+    companion object {
+        fun forExportedLibraries(configuration: CompilerConfiguration): FeaturedLibrariesReporter = ExportedLibrariesReporter(configuration)
+        fun forSourceLibraries(configuration: CompilerConfiguration): FeaturedLibrariesReporter = SourceLibrariesReporter(configuration)
+        fun forCoveredLibraries(configuration: CompilerConfiguration): FeaturedLibrariesReporter = CoveredLibraryReporter(configuration)
+    }
+}
+
+private fun getFeaturedLibraries(
+    featuredLibraries: List<String>,
+    resolvedLibraries: KonanLibraryResolveResult,
+    resolver: SearchPathResolver,
+    reporter: FeaturedLibrariesReporter
+) : List<KonanLibrary> {
+    val featuredLibraryFiles = featuredLibraries.toUnresolvedLibraries.map { resolver.resolve(it).libraryFile }.toSet()
+    val remainingFeaturedLibraries = featuredLibraryFiles.toMutableSet()
+
+    val result = mutableListOf<KonanLibrary>()
+
+    val libraries = resolvedLibraries.getFullList(null)
+
+    for (library in libraries) {
+        val libraryFile = library.libraryFile
+        if (libraryFile in featuredLibraryFiles) {
+            remainingFeaturedLibraries -= libraryFile
+            if (library.isInterop || library.isDefault) {
+                reporter.reportIllegalKind(library)
```

Comment:
IIRC, coverage was allowed for default libraries before. Is it forbidden intentionally here?

Reply:
I've discussed this point with @sbogolepov and he said that default libraries are not intended to be covered.

Reply:
Hmm, coverage for stdlib may make sense.

Reply:
Fixed